### ISSUE TITLE
Supporting alternatives to Google Drive

### ISSUE THREAD

**Issue by fklemme** [chilli-axe/mpc-autofill#39]
While Google Drive is a nice interface in a distributed setup, for the time being, it would be helpful to have some alternative for the local executing of the web application. Without a collection of Google Drives available, it would be helpful if own files could be offered easily without the need for a (large, paid) Google Drive. I could imagine having a local folder as an additional source of images or some even some other web-space. While I would like to help and contribute in developing such a feature, I think this topic requires the expertise of @ndepaola to judge what is doable and reasonable. Maybe we can do some brainstorming here. I was thinking of a few possibilities:

1. Support a local drive. This seems to be the most obvious solution but might be tricky to implement. First, Django needs to be aware of the folder to scan it, then it needs to be served via static files to be displayed. And, last but not least, it would need to be added to the order.csv in a way that the files can be copied by the `autofill.exe` for MPC upload.
2. Support a common website directory listing, e.g., as generated by Apache (see https://kruecke.net/files/, for example). Then, a local folder (or any other source) could be published using a simple Apache. Django would need to be able to search the directory listing and index all images. Since the files are downloadable via URL, this might be more portable and less error-prone in configuration, also for the autofill client.
3. Based on the previous, I could imagine building a small php script (or even a small django app) to do the indexing and offer all information needed by MPCAutofill directly as JSON or similar. This could simplify the changes to MPCAutofill and only a small extra installation to the offering website is needed.

Anyway, I think all methods would require to add an additional field to the `drives.csv` and go from there.

What do you think? Is this something worth targeting? Doable in reasonable effort? Or is this rather a different project altogether?

**Comment by avafloww** [comment]
This is actually something I was investigating/starting to work on setting up myself, at least for #2 (or a variation of it - i.e. using nginx `autoindex_format json` output). I'll poke around at it later; putting this comment here to remind myself 😄

**Comment by ndepaola** [comment]
i agree & think that local file support is the future of this project - was planning on having a crack at this week.

imo the way to do this while minimising structural changes to the project is to index local files within the current three elasticsearch indexes, which means modifying those ORM tables to support local files and gdrive files. it might also make sense to have a user interface for linking folders to mpc autofill and a second database updater script which crawls local files, accessible with a button press rather than requiring the command line. in that instance, we'd want a django for local mode vs hosted and only enable local file support in local mode.

google drive is baked into mpc autofill fairly deeply so this'd involve a bit of frontend work to handle local files differently to gdrive images as well.

storing local file paths in xml and using them when autofilling should be fairly straightforward.

thoughts?

**Comment by ndepaola** [comment]
that's a very good point - i dig the idea of optionally pointing the docker image at a single directory to index (and we could set this up such that the docker config modifies a django setting, meaning the project is still usable without docker). it may also be possible to run the autofill script from within the docker image, eliminating the need for downloading the images from the docker image back to the host file system. 

definitely planning on retaining google drive functionality alongside local file support!

re: `client_secrets` - the file is only required by the drive crawler management script (`update_database.py`). once drive files are indexed, the frontend retrieves thumbnail images and supports downloading the full res images without needing to authenticate with google.

**Comment by fklemme** [comment]
Having one folder configured in an `env` might be just it. Django can read that, docker-compose can read that (to then mount and re-route django to, e.g., `/cards`. Still, it would be good if the serve the images trough staticfiles for portability rather than letting the HTML point to file on the local filesystem. I'm not 100% sure to do this best in a development setup, though.

Oh man, I never even considered putting the client into docker as well. :heart_eyes: Got to look into that. The tricky part would be that docker doesn't offer a GUI natively. So maybe we need the user to enter MPC credentials, run Chrome headless and store the uploaded project (as offered by MPC). Otherwise the user might need something like VNC client to display the Chrome running in Docker. But I will have a look what options are available.

**Comment by ljrobison** [comment]
I am only a novice self taught programmer so I don't know much of anything I can offer to help, but if you guys could local folder support, or even a network drive.. that would be absolutely amazing. Now I just need to figure out how to get this running as a docker on my unRAID server haha.

**Comment by avafloww** [comment]
![Refactored the autofill client to support both drive and regular HTTP links (video link) - now to implement support for it into the site! 😄](https://youtu.be/piI_EMZVgZs)

(Also, the URLs shown in the xml file in the video are internal to my LAN only, so don't bother typing them out)

**Comment by ndepaola** [comment]
in terms of the local tool, i think the way to go will be storing local files' paths in the `<id>` tag in xml - it'd be difficult to determine whether an image is from google drive or is locally stored from xml without changing the xml schema. adding support for this in my local tool rewrite branch https://github.com/chilli-axe/mpc-autofill/tree/local-tool-rewrite remote isn't up to date atm but i'll push my changes shortly.

**Comment by ndepaola** [comment]
> 
> 
> ![Refactored the autofill client to support both drive and regular HTTP links (video link) - now to implement support for it into the site! 😄](https://camo.githubusercontent.com/09db9e90930fee4d000d6b530843d93761454dfa6198c490d37ce820324bafd1/68747470733a2f2f796f7574752e62652f7069495f454d5a56675a73)
> 
> (Also, the URLs shown in the xml file in the video are internal to my LAN only, so don't bother typing them out)
> 
> (edit, in case the URL above is broken because thanks GitHub: https://www.youtube.com/watch?v=piI_EMZVgZs)

nice work! the local tool in master is a truly horrible piece of code so i'm sorry about that but i'm improving i swear 😅

**Comment by fklemme** [comment]
>         <card>
>             <id>G:\Google Drive\Chilli_Axe's MTG Renders\0. White\Academy Rector.png</id>
>             <slots>1,2,0</slots>
>             <name>Academy Rector.png</name>
>             <query>academy rector</query>
>         </card>

Will the client still "download" the files to a cards sub-folder? Please keep in mind that in practice, people might have deep folder structures containing duplicate filenames. Thus, some kind of ID should still be appended to the filename. (Maybe just a hash, in case of path?) Or will you be extending the schema in your rewrite branch?

**Comment by ndepaola** [comment]
in my branch, the logic is now:

parsing and validation step:
* if the contents of the `<id>` tag points to a valid file, use this as the image's file path
* otherwise, if the image doesn't have a file name from xml (e.g. the common cardback), the image's file name is retrieved from google scripts api
* then, the system uses the image's file name and looks in `/cards` to see if the image exists without the drive ID in parentheses - this has the potential to cause file name collisions, but is what allows the tool to work when using the `Download All` button in the web app and moving those files to `/cards` - if a file exists at that path, use this as the image's path
* finally, set the image's path to point at `/cards` using the file's gdrive ID in parentheses

downloader threads:
* if no file exists at the file's path, attempt to download it from google drive based on the image's ID
* if that fails, report on being unable to download the image and skip inserting it

so local files are uploaded to mpc directly from the paths they already exist at (they aren't copied to `/cards`)

not sure if i explained that clearly sorry but hopefully it makes sense!

**Comment by fklemme** [comment]
Sorry, I'm not that familiar with the codebase: Wouldn't it be worth the effort to add another optional tag (e.g., `path`) to be explicit and avoid future confusions? Would that actually require many changes?

**Comment by ndepaola** [comment]
you're probably right yea - i suppose it's not a problem if some cards don't have a `<path>` tag! will think about it more but i'll probably end up doing this

**Comment by fklemme** [comment]
I appreciate your consideration of backwards compatibility. However, now that you're rewriting the client anyway would be the best time to fix these things. Also, I believe most people also won't get into trouble with a change because they can a) still grab an older release or b) will use your web application anyway, in which we can apply the same changes. So I won't be too defensive when aiming for a good and future-oriented change. :)

I'm also working on some ideas to bundle the client with docker. This might further reduce the chance of picking incompatible tool versions. I might come back to you with some Django-related questions, because I'm considering to add a button to launch the client directly from the web interface, and I'm not too familiar with Django / web development.

**Comment by avafloww** [comment]
Oh dear, looks like there's several cooks in the kitchen on this 😅

I'll have a look later at adding HTTP download support to the rewritten autofill tool. I've also started on abstracting sources to allow for different parameters for different source types (like drive id and drive link, or URL, or local file path) on the web side of things, but I'm not 100% sure I'm going about it the right way yet. I'll try to get some code up within the next few days if you wanted to take a look!

**Comment by chilli-axe** [comment]
![Screenshot (45)](https://user-images.githubusercontent.com/91188799/135756115-546099f5-73eb-455a-8212-27f57584725b.png)

getting there with this feature! hoping to have a pr up before too long

**Comment by fklemme** [comment]
Wow, that's a lot of new code! :smile:
Just one question so far. With adding local files to the static files like this:
```
LOCAL_FILE_INDEX = r""  # for example: r"C:\Users\John Doe\Desktop\MPC Cards"
# [...]
STATICFILES_DIRS = [
    os.path.normpath(os.path.join(BASE_DIR, "cardpicker/static")),
    os.path.normpath(LOCAL_FILE_INDEX),
]
```
Does this mean, when I call `python3 manage.py collectstatic`, all cards will be _copied_?

**Comment by chilli-axe** [comment]
yes - it works fine in development but will require some more thought for dockerising since you're serving static files with nginx. a few ideas:

* ideally, we want `collectstatic` to collect all files that aren't in `LOCAL_FILE_INDEX` to the django static directory, then nginx should serve that directory as well as `LOCAL_FILE_INDEX` on `/static` - might that be possible? for what it's worth, `collectstatic` has an optional argument to ignore files matching a pattern but i can't seem to force it to ignore my local file index.
* it might make sense to configure `LOCAL_FILE_INDEX` in a text file in the base directory so you don't need to modify django settings to configure the local index - this might open up the possibility of copying the value from the configuration file into django settings *after* running `collectstatic`?
* lastly, the `update_database` script could create low resolution thumbnails of all images in `LOCAL_FILE_INDEX` and django could serve these rather than the full res images, then copying the thumbnails with `collectstatic` would be less costly than copying the full resolution images. might still require a non-trivial amount of storage space though and i haven't tested how long creating thumbnails in this way might take.

open to any suggestions on this!

**Comment by fklemme** [comment]
I was also thinking of the first option intuitively. It should be very easy to implement in Docker, so we should just try this one first. The second option would work just as well but I don't think it's necessary. The third option sounds interesting for performance reasons. But we should first try the simple way before we blindly optimize for something that might not be necessary in the first place.

I will check out the branch soon and give it a try. Then I'll also see if there are other things that we'll need to consider.

**Comment by rsullivan00** [comment]
For people that already have directories with only the images they want to use, and don't want to have to use the web interface, I hacked together a script. It can load front images from one directory, and back images from another.

It does require some editing, and manually setting the `slot` of backs of cards after the XML is generated.

https://gist.github.com/rsullivan00/df968d764101a84244b4b1a06caecf79
